Source/Blazorise/Components/Bar/BarDropdown.razor.cs

```diff
@@ -265,6 +268,13 @@ public bool Visible
 
             state = state with { Visible = value };
 
+            if ( value )
+            {
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
```

Comment:
Hm, not liking this code smell. Can we use `ExecuteAfterRender`?

Or make is as async discard https://stackoverflow.com/questions/42838362/ignoring-the-return-value-from-an-async-task-method?

Reply:
The code smell is already because of the architecture of the component in my opinion that makes use of the setter like this. Also we already have another async call there, VisibleChanged.InvokeAsync, that should also not be appropriate.

I'd prefer to do ParametersSet and check for visible changed as we've seen recommended before, if you really want to change.

Reply:
> Or make is as async discard https://stackoverflow.com/questions/42838362/ignoring-the-return-value-from-an-async-task-method?

I don't see the point of this? It is equivalent to what we have. 
Let's not mistake a compiler warning not showing for something, making it right suddently.

Reply:
Ok, change it to ParametersSet. We can start slowly with that change whenever we work on some component.

Reply:
The point is to be consistent, in my opinion I would leave it for consistency in order to not have mixed ways to do things. 
If we chose this "architecture" for this component, we stick to it in my opinion. If you still prefer that I move it, I'm ok with moving it.

Reply:
I agree to move it. We need to do it anyway. I'm just saying that by doing it peace by peace, we're making it easier for ourselves in the future, instead of going all-in on all components at once.

Reply:
Hopefully done. Please retest on your side.

Reply:
That does seem like it would work more as I personally would expect, with a private variable tracking the visible state and listening to the `Visible` Parameter. But I need to take a look better at it, because I'm not entirely convinced it's synced correctly. Also I believe you forgot to trigger `VisibleChanged`.

Also I'm still puzzled why we keep creating a new object reference for the state each time. Why not just use a class if we are not using the record semantics apparently? 
So basically just doing : `state.Visible = trueOrFalse`?

Reply:
The record is used because we want the component children to keep track of the state changes. The class is checked by reference, and the record is checked by the values.

Reply:
Ok gotcha, so it's for the automatic change detection. Could it no be a struct then?

Reply:
I honestly think it should. Let me find docs

Reply:
Actually, What if it's being triggered because you're newing up an object instead of the equality factor?

Reply:
You could argue that you could implement some kind of abstraction where you'd have a State Base Class where you have equality as a mandatory override. 
And you could have the best of both worlds, since you don't need the immutability feature and wouldn't need to pay the cost to new up an object every single time.

Reply:
**PS:** There's also **record struct**

Reply:
It might indeed be because of the newing of the record object. But there is another neat record feature. The `with` keywoard. It is a lot easier to clone an object that way than to have a separate `Clone()` method. It's all handled by the compiler.

Reply:
You didn't read everything... :) It's more ceremony indeed. But consumers of your library shouldn't have to pay the additional memory cost because you want to write less code if you can avoid it.

- Either a source generated class **OR**
- a record struct
could possibly at least remove the price we're paying with all the **With**. Which I honestly can't tell you if that price is huge or not in the grand scheme of things.

> You could argue that you could implement some kind of abstraction where you'd have a State Base Class where you have equality as a mandatory override.
And you could have the best of both worlds, since you don't need the immutability feature and wouldn't need to pay the cost to newing up an object every single time. Just more cerimony of course. But with source code generator that could be mitigated.

Reply:
I will need to investigate more about `record struct`.

But isn't regular regular struct always copied when assigned to another variable? Which potentially means cascaded struct could be copied on each rerender. 🤔 Which would have even more impact on the performance. With the current record, we at least try to clone it only when needed.

Reply:
As far as my understanding goes, record struct is a struct with record semantics. 
This means that you should get the benefits of a struct. 
Which means it should act as any value type, like an int for instance. So it should work as any other instance value type member of that class as far as I understand and should remove the allocated heap memory out of the equation and all the objects being created each time and having to be garbage collected.
So I don't understand your reasoning on it having more impact.

Reply:
But again, we can just open a separate issue and investigate that as part of the separate issue. And use some memory profile tooling to actually take concrete evidence.

Reply:
https://learn.microsoft.com/en-us/dotnet/csharp/advanced-topics/performance/
![image](https://github.com/Megabit/Blazorise/assets/22283161/61024a48-bc20-4539-843b-ed2f78e4918d)

Reply:
Yeah, I'm aware of the difference between class and struct.

---

I quickly updated and converted `record` to `record struct`. Here are the results:

![image](https://github.com/Megabit/Blazorise/assets/900302/debdecee-e676-4ff8-8896-b4c818c2d324)

![image](https://github.com/Megabit/Blazorise/assets/900302/34a25077-7e80-43d3-a979-7cae038b2693)

So a lot of our internal logic would break with value based types.

Reply:
I see. We could still do some tricks to check whether the state really exists instead of null checks, seems like it would be an easy fix. But anyway if you don't want to investigate this at all it's fine. Resolving.